Fix HeMIS related bugs

andreanne-lemay opened the pull request:
`task` attribute is missing from the adaptive loader

The adaptive loader was not adapted to multiclass training.

Changes include:
- change None to 'None' since it is not a supported data type for h5py
- store all GT 
- `target_lst` became `target_contrast`. `target_contrast` indicates which GT to use (from which contrast)
- `roi_lst` became `roi_contrast`. `roi_contrast` indicates which ROI to use (from which contrast)

Old behavior: when `target_lst`had a length > 1, all GTs were used (the same way we store multiclass GT). For example, `target_lst`: ["T1w", "T2w"] would lead to a GT of size 2xHxW, but the prediction of the model would be 1xHxW for single class models. On multichannel, we want to pick a single GT for each class (from one of the contrast) to have the same behavior we have on the regular loader.

The adaptive loader saves a single dataset file (.hdf5) so the same dataset is used for validation, training and testing. Changes introduced in this PR fixes this behavior.

Remaining issues with HeMIS for future PRs:
- The adaptive loader works only for HeMIS (2D, multichannel)
- We can only test on datasets with all input channels

andreanne-lemay committed:
adaptative: add task attribute

charleygros reviewed (approved):
:+1:

andreanne-lemay committed:
change None to 'None'

andreanne-lemay committed:
adaptative loader for multiclass

andreanne-lemay committed:
fix typo

andreanne-lemay commented on ivadomed/loader/adaptative.py:
```diff
@@ -350,16 +350,16 @@ def _load_filenames(self):
-                    gt_volume = []
-                else:
-                    gt_volume.append((slice_seg_pair["gt"][0] * 255).astype(np.uint8) / 255.)
+                gt_slice_list = []
```
Store all gt (not only the one in index 0)

andreanne-lemay commented on ivadomed/loader/adaptative.py:
```diff
@@ -608,19 +611,19 @@ def __getitem__(self, index):
-            gt_img.append(gt_data)
+        gt_metadata = []
+        for n_gt in range(gt_img.shape[0]):
             gt_metadata.append(imed_loader_utils.SampleMetadata({key: value for key, value in
```
Store as many gt metadata as there are gts

andreanne-lemay committed:
add comment

andreanne-lemay committed:
change target_lst to target_contrast

andreanne-lemay committed:
add documentation on HeMIS params

andreanne-lemay committed:
change roi_lst to roi_contrast

andreanne-lemay committed:
update test with roi_contrast

andreanne-lemay committed:
Merge branch 'master' into al/hemis_bug_fix

charleygros commented on docs/source/configuration_file.rst:
```diff
@@ -323,6 +323,12 @@ HeMISUnet (Optional)
    modified with the exponent ``missing_probability_growth``.
+- ``contrasts``: List. List of contrasts used as input.
+- ``ram``. Bool. Indicates whether to save the data in ram or not.
+- ``hdf5_path``: Str. Path to hdf5 files containing the dataset.
```
```suggestion
- ``hdf5_path``: Str. Path to hdf5 file containing the dataset.
```

charleygros commented on docs/source/configuration_file.rst:
```diff
@@ -323,6 +323,12 @@ HeMISUnet (Optional)
+- ``contrasts``: List. List of contrasts used as input.
+- ``ram``. Bool. Indicates whether to save the data in ram or not.
+- ``hdf5_path``: Str. Path to hdf5 files containing the dataset.
+- ``csv_path``: Str. Path to dataframe.
```
```suggestion
- ``csv_path``: Str. Path to dataframe (csv file).
```

charleygros commented on docs/source/configuration_file.rst:
```diff
@@ -323,6 +323,12 @@ HeMISUnet (Optional)
+- ``ram``. Bool. Indicates whether to save the data in ram or not.
+- ``hdf5_path``: Str. Path to hdf5 files containing the dataset.
+- ``csv_path``: Str. Path to dataframe.
+- ``target_contrast``: Str. Contrast label used for ground truth.
```
```suggestion
- ``target_contrast``: Str. Contrast whom the label file is used as ground truth.
```

charleygros commented on docs/source/configuration_file.rst:
```diff
@@ -323,6 +323,12 @@ HeMISUnet (Optional)
+- ``hdf5_path``: Str. Path to hdf5 files containing the dataset.
+- ``csv_path``: Str. Path to dataframe.
+- ``target_contrast``: Str. Contrast label used for ground truth.
+- `` Contrast label used for ROI cropping.``: Str. Contrast label used for ROI cropping.
```
```suggestion
- `` Contrast label used for ROI cropping.``: Str. Contrast whom the label file is used for ROI cropping.
```

charleygros reviewed (approved):
Noiiice solution! Like it

@leandervaneekelen : would you mind running HeMIS on your dataset / application using this branch? Many thanks!

andreanne-lemay committed:
Update docs/source/configuration_file.rst

Co-authored-by: charleygros <charley.gros@gmail.com>

andreanne-lemay committed:
Update docs/source/configuration_file.rst

Co-authored-by: charleygros <charley.gros@gmail.com>

andreanne-lemay committed:
Update docs/source/configuration_file.rst

Co-authored-by: charleygros <charley.gros@gmail.com>

andreanne-lemay committed:
Update docs/source/configuration_file.rst

Co-authored-by: charleygros <charley.gros@gmail.com>

andreanne-lemay commented on docs/source/configuration_file.rst:
```diff
@@ -323,6 +323,12 @@ HeMISUnet (Optional)
+- ``hdf5_path``: Str. Path to hdf5 file containing the dataset.
+- ``csv_path``: Str. Path to dataframe (csv file).
+- ``target_contrast``: Str. Contrast whom the label file is used as ground truth.
+- `` Contrast label used for ROI cropping.``: Str. Contrast whom the label file is used for ROI cropping.
```
```suggestion
- ``roi_contrast``: Str. Contrast whom the label file is used for ROI cropping.
```

andreanne-lemay committed:
Update docs/source/configuration_file.rst

andreanne-lemay committed:
Merge branch 'master' into al/hemis_bug_fix

leandervaneekelen commented:
Looks good! 

During testing, when I run the code, `model_params` in [`ivadomed/loader.py`](https://github.com/ivadomed/ivadomed/blob/874d258c7cc4a8a715063d5e13381084b6908e52/ivadomed/loader/loader.py#L75), still contains the keys 'roi_lst' and 'target_contrast' though, so further down the line it fails - maybe some missed replacements?

andreanne-lemay committed:
roi_contrast to roi_lst

andreanne-lemay commented:
Nice catch, I didn't test on testing command! It seems like there are deeper issues.
During testing, we need to undo the transforms. It seems like the cropping parameters are not retrieved / saved leading to an error here: https://github.com/ivadomed/ivadomed/blob/master/ivadomed/transforms.py#L413.

andreanne-lemay committed:
Save crop params

andreanne-lemay committed:
don't discard subjects that misses some GT for multichannel

andreanne-lemay committed:
minor fix in loader

andreanne-lemay committed:
Use bids dataset in adaptive (reduce code duplication)

andreanne-lemay committed:
create one hdf5 file per dataset type

andreanne-lemay committed:
update HeMIS and adaptive tests

andreanne-lemay committed:
change name hdf5 file in test

andreanne-lemay committed:
change name hdf5 file in test

andreanne-lemay committed:
Merge branch 'master' into al/hemis_bug_fix

leandervaneekelen committed:
Make HDF5Dataset class return all modalities

leandervaneekelen committed:
Fix shape error for multi-contrast HeMIS usage

andreanne-lemay commented on docs/source/configuration_file.rst:
```diff
@@ -336,6 +336,12 @@ HeMISUnet (Optional)
 -  ``missing_probability_growth``: Float. Controls missing probability
    growth at each epoch: at each epoch, the ``missing_probability`` is
    modified with the exponent ``missing_probability_growth``.
+- ``contrasts``: List. List of contrasts used as input.
```
Add missing doc

andreanne-lemay commented on ivadomed/config/config_spineGeHemis.json:
```diff
@@ -73,8 +73,8 @@
         "csv_path": "/home/andreanne/Documents/dataset/HeMIS.csv",
-        "target_lst": ["T2w"],
-        "roi_lst": null
+        "target_contrast": "T2w",
```
Now, only one contrast can be selected as GT for `target_contrast` and `roi_contrast`.

andreanne-lemay commented on ivadomed/loader/adaptative.py:
```diff
@@ -25,7 +25,7 @@ class Dataframe:
         path (str): Dataframe path.
         target_suffix (list of str): List of suffix of targetted structures.
-        roi_suffix (str): List of suffix of ROI masks.
+        roi_suffix (list): List of suffix of ROI masks.
```
Fix error in doc string

andreanne-lemay commented on ivadomed/loader/adaptative.py:
```diff
@@ -175,122 +175,47 @@ class Bids_to_hdf5:
         slice_filter_fn (SliceFilter): Object that filters slices according to their content.
     """
 
-    def __init__(self, root_dir, subject_lst, target_suffix, contrast_lst, hdf5_name, contrast_balance=None,
```
Remove code duplication with [BidsDataset](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/loader/loader.py#L750). Most of the code was copied from there with minor additions.

andreanne-lemay commented on ivadomed/loader/adaptative.py:
```diff
@@ -175,122 +175,47 @@ class Bids_to_hdf5:
+            subject_id = filename.split("_")[0]
+            list_patients.append(subject_id)
+            filename_pair = list(filename_pair)
+            filename_pair.insert(0, subject_id)
```
Add subject id to filename_pair

andreanne-lemay commented on ivadomed/loader/adaptative.py:
```diff
@@ -371,100 +298,112 @@ def _load_filenames(self):
+            for slice_metadata in slice_seg_pair['input_metadata']:
+                contrasts.add(slice_metadata['contrast'])
             # Creating datasets and metadata
-            contrast = input_metadata['contrast']
```
Loop across all contrasts instead of one contrast from the first input_metadata.

andreanne-lemay commented on ivadomed/loader/adaptative.py:
```diff
@@ -492,38 +431,45 @@ class HDF5Dataset:
-                                     target_suffix=target_suffix,
-                                     roi_params=self.roi_params,
-                                     contrast_lst=self.cst_lst,
+            dataset = imed_loader.BidsDataset(root_dir=root_dir,
```
Use BidsDataset instead of duplicating code in Bids_to_hdf5

andreanne-lemay commented on ivadomed/loader/adaptative.py:
```diff
@@ -590,72 +536,75 @@ def __getitem__(self, index):
             else:
-                input_tensor = self.hdf5_file[line[ct]][line['Slices']] * missing_modalities[i]
+                tensor = self.hdf5_file[line[ct]][line['Slices']] * missing_modalities[i]
+            input_tensors.append(tensor)
```
Append all modalities into a list

andreanne-lemay committed:
fix conflicts with master

andreanne-lemay committed:
fix merging with master

andreanne-lemay committed:
test_HeMIS: fix bug

andreanne-lemay committed:
loader: keep only subjects with GT for training

andreanne-lemay committed:
adaptive: store slices by contrast

andreanne-lemay commented on ivadomed/loader/adaptative.py:
```diff
@@ -372,99 +303,108 @@ def _load_filenames(self):
-            # dataset metadata
-            grp[key].attrs['gt_filenames'] = input_metadata['gt_filenames']
-            grp[key].attrs['data_type'] = gt_metadata['data_type']
+            for c_idx, contrast in enumerate(self.contrasts):
```
Input data have 2 or more contrasts -> loop across all contrasts

andreanne-lemay committed:
Bids_to_hdf5: make contrasts mandatory param

leandervaneekelen committed:
Revert "Fix shape error for multi-contrast HeMIS usage"

This reverts commit 44d2bf04a71bc4a0ec8a24ed1132de6d3e0f2d6a.

leandervaneekelen commented on ivadomed/training.py:
```diff
@@ -165,7 +165,6 @@ def train(model_params, dataset_train, dataset_val, training_params, log_directo
         for i, batch in enumerate(train_loader):
             # GET SAMPLES
             if model_params["name"] == "HeMISUnet":
-                batch["input"] = batch["input"].squeeze() if batch["input"].shape[1] > 1 else batch["input"]
```
I reverted an earlier commit that changed the shape of `DataLoader`'s `batch["input"]` for HeMIS; I mistakenly made this commit when using an improperly formatted dataset. With correctly formatted data, this line is superfluous and can actually cause errors, e.g. in #594 where the `filter_absent_mask` can cause `batch["input"]`'s length to be smaller than the batch size.

Apologies for this ugly revert-commit, I couldn't figure out how to do this neatly with a `git rebase` :sweat:

leandervaneekelen committed:
Save entire crop params dict into HDF5 metadata attributes

HDF5 doesn't allow dictionaries to be assigned as metadata attributes,
so as a work-around, key:value pairs from the dict are now stored as
grp[key].attrs["crop_params/<key] = value, and later retrieved properly
properly in `HDF5Dataset.__getitem__`.

leandervaneekelen committed:
Save preresample_shape in HDF5 file

gisellemartel commented:
As @mariehbourget pointed out, this PR actually seems to already address a new issue #703.
